src/durable-cache/src/lib.rs

```diff
@@ -293,6 +300,43 @@ impl<C: DurableCacheCodec> DurableCache<C> {
             }
         }
     }
+
+    async fn compare_and_append<I>(
+        &mut self,
+        updates: I,
+        expected_upper: u64,
```

Comment:
Personally, I'd do something like `fn compare_and_append(.., write_ts: u64, new_upper: u64)` (where write_ts is used as the ts in all the updates) and then in the impl do `let expected_upper = write_ts`. That allows you to do the downgrade_since to `write_ts`. We don't need to downgrade to "upper-1" so much as we need to downgrade to "at or past the most recent write".

That said, what you have here works. I just think this is slightly more semantically obvious.

Reply:
That's a good idea, just updated the code to look more like this.